zebra-state/src/service/chain_tip.rs

```diff
@@ -320,6 +320,31 @@ impl ChainTipChange {
         Ok(action)
     }
 
+    /// Returns:
+    /// - `Some(`[`TipAction`]`)` if there has been a change since the last time the method was called.
+    /// - `None` if there has been no change.
+    ///
+    /// See [`wait_for_tip_change`] for details.
+    pub fn last_tip_change(&mut self) -> Option<TipAction> {
+        // Obtain the tip block.
+        match self.receiver.borrow().as_ref() {
+            Some(block) => {
+                // Ignore an unchanged tip.
+                if Some(block.hash) == self.last_change_hash {
+                    return None;
+                }
+
+                let action = self.action(block.clone());
+
+                self.last_change_hash = Some(block.hash);
+
+                Some(action)
+            }
+
+            None => None,
+        }
```

Comment:
Optional suggestion: You can remove two levels of indentation by using the try operator (`?`), but that would require adding an extra binding:

```suggestion
        let block_guard = self.receiver.borrow();
        let block = block_guard.as_ref()?;
        
        // Ignore an unchanged tip.
        if Some(block.hash) == self.last_change_hash {
            return None;
        }

        let action = self.action(block.clone());

        self.last_change_hash = Some(block.hash);

        Some(action)
```

Reply:
I was trying the `?` operator, but I couldn't get it to work. Thanks.

Reply:
No problem :D Yeah, the tricky part is satisfying the borrow checker. `self.receiver.borrow()` returns a wrapper type that `Deref`s into its inner type (i.e., `Option<_>`), so you can call `.as_ref()` directly. The problem is that chaining calls makes Rust drop intermediate types as soon as possible, so the wrapper is dropped because it's not stored anywhere. The chain of events is (at least AFAICT):

1. The wrapper type is obtained and stored in a temporary;
2. That temporary is then borrowed and dereferenced to get a borrow to the inner type, i.e. `&Option<T>`;
3. That borrow is used to call `fn as_ref(&self) -> Option<&T>`, so now we have a result that depends on the borrow;
4. The wrapper type is dropped;
5. The result is stored.

The problem is in steps 4 and 5, because we still have inside the result a borrow to the temporary, even though it was dropped. So the Rust compiler fails to compile the code. By creating the `blocks_guard` binding we make sure that it is only dropped at the end of the scope.

Reply:
Thanks for the explanation!